### PR TITLE
Add debugging options for terraform tests in CI

### PR DESCRIPTION
**Pull request by dominikschubert**
Trying to find the issues that cause the terrarform tests to fail recently.

A few notes (will be updated as the PR progresses):
- changing the AWS region fixes some of these issues. Not all AWS AMIs in EC2 are available in all regions, so for example setting region to "us-west-2" fixes one error in the EC2 tests complaining about an AMI not existing. This indicates that the recent moto + store changes might be related somehow to these issues.

Changes in this PR:

- GitHub action has more control over the refs we checkout when using workflow_dispatch (i.e. manually initiated runs).
- GitHub action stores debug logs for each test and uploads them on failure ( => might need to be adjusted a bit since this produces quite a lot of output). They're only uploaded when the tests fail.

### CODE REVIEW

**Comment by bblommers** [comment]
The tests that Moto uses fail as well, and as far as I can tell, these failures were introduced when Terraform released version 1.3.0.
Terraform 1.3.1 fixed some of the issues, notably `Failed to serialize resource instance in state`, but not all - the `Failed state verification` failures still appear with 1.3.1.

**Comment by bblommers** [comment]
I couldn't find anyone else complaining about this, so I've opened an issue here: https://github.com/hashicorp/terraform-provider-aws/issues/27049

**Comment by dominikschubert** [comment]
Yeah I kinda stumbled upon https://github.com/hashicorp/terraform-provider-aws/issues/26962 (and https://github.com/hashicorp/terraform/issues/31851) today as well, but as you said there are still some other issues left like all the missing IDs unfortunately.

Let's hope they'll move as quickly as they did for the "serialize resource" issue. :smile:

**Comment by thrau** [comment]
@dominikschubert i rebased the branch which now includes #6962 

i think the dispatch parameters are definitely worth keeping. do you want to clean this up and then merge it as an enhancement to the action?

**Comment by dominikschubert** [comment]
> i think the dispatch parameters are definitely worth keeping. do you want to clean this up and then merge it as an enhancement to the action?

@thrau  I think the additional logs artifacts might also still be useful when debugging CI failures in terraform. It will only store logs of the failed test files anyway. wdyt? :thinking: 
Otherwise I can ofc also just remove them :grin:

**Comment by alexrashed** [comment]
FYI: I think current test failures are caused by changes in the tests ([they check for EC2 VPC's enable_network_address_usage_metrics, which are not yet supported by LocalStack](https://github.com/hashicorp/terraform-provider-aws/commit/8dab9f1892be61e9441e9a582a052016a3750ac8)).

File: .github/workflows/terraform-tests.yml [hunk 1]
```diff
@@ -56,6 +68,8 @@ jobs:
       uses: actions/checkout@v3
       with:
         repository: hashicorp/terraform-provider-aws
+        ref: v4.31.0
+#        ref: ${{ github.event.inputs.terraform-ref }}
```
Comment (alexrashed) [inline .github/workflows/terraform-tests.yml:72]: This seems to be a leftover.
A lot of the tests are currently failing due to changes in the tests (they check for a new attribute in EC2 VPCs which is not yet supported by LocalStack). Therefore using a pinned version would be a good idea for now. But the input should still be used. So the default behavior should be that the terraform-ref input is used when set, but it defaults to the latest ref which is known to be working (i.e. the pinned version).

```suggestion
        ref: ${{ github.event.inputs.terraform-ref || "v4.31.0"}}
```

**Review by alexrashed — CHANGES_REQUESTED** [review]
Thanks for tackling this! The terraform tests are causing some issues lately, so this will come in very handy. 🔝 
The implementation looks good to me, besides one small leftover. 🚀

**Review by alexrashed — APPROVED** [review]

**Pull request merged**